sched: make CycleState's Read()/Write()/Delete() thread-safe

Huang-Wei opened the pull request:
#### What type of PR is this?

/kind bug
/sig scheduling

#### What this PR does / why we need it:

This is a followup of the TBD item in https://github.com/kubernetes/kubernetes/issues/95887#issuecomment-722187136.

Scheduler used and exposed a map called `CycleState` to store transient data within one scheduling cycle. This struct is usually keyed with a particular plugin name + phase (PreFilter or PreScore), and valued with the concrete data. Without this PR, this struct is:

1. written without locking in PreFilter / PreScore phases
2. read without locking in Filter / Score phases

Behavior 2 should be adjusted to acquire read lock before using it as we shouldn't make that assumption for out-of-tree plugins - which is how the racing condition occured in #95887. (although in-tree plugins don't write data to `CycleState`)

In terms of behavior 1, although in-tree plugins use Write() in sequential phase, and hence _can_ be lock-free. But again, to eventually remove misleading Lock()/Unlock() functions in CycleState, we decide to make no assumption which phase it can/should be used, in other words, add lock inside Write() as well.

Note that the racing pattern this PR solves is **multiple plugins run in parallel** to compete for `CycleSatate`. Another racing pattern is **the same plugin runs in parallel (on different nodes)**, which can be resolved by the plugin's implementation itself (check #96777 for more details).

#### Which issue(s) this PR fixes:

Fixes #95887

#### Special notes for your reviewer:

I didn't see an obvious performance downgrade according to the bench diff:

<details>
  <summary>SchedulingBasic/5000Nodes</summary>
  
  ```
  +--------------------------+---------------------------+----------+--------+---------+---------+---------+
  |          METRIC          |           GROUP           | QUANTILE |  UNIT  |   OLD   |   NEW   |  DIFF   |
  +--------------------------+---------------------------+----------+--------+---------+---------+---------+
  | SchedulingThroughput     | SchedulingBasic/5000Nodes | Average  | pods/s |  166.67 |  162.56 | -2.47%  |
  | SchedulingThroughput     | SchedulingBasic/5000Nodes | Perc50   | pods/s |  178.56 |  184.89 | +3.55%  |
  | SchedulingThroughput     | SchedulingBasic/5000Nodes | Perc90   | pods/s |  208.00 |  202.30 | -2.74%  |
  | SchedulingThroughput     | SchedulingBasic/5000Nodes | Perc99   | pods/s |  209.00 |  202.50 | -3.11%  |
  | scheduler_e2e_scheduling | SchedulingBasic/5000Nodes | Average  | ms     |   26.75 |   25.30 | -5.41%  |
  | scheduler_e2e_scheduling | SchedulingBasic/5000Nodes | Perc50   | ms     |   12.53 |   12.52 | -0.09%  |
  | scheduler_e2e_scheduling | SchedulingBasic/5000Nodes | Perc90   | ms     |   29.98 |   30.88 | +3.01%  |
  | scheduler_e2e_scheduling | SchedulingBasic/5000Nodes | Perc99   | ms     |  524.08 |  446.84 | -14.74% |
  | scheduler_pod_scheduling | SchedulingBasic/5000Nodes | Average  | ms     | 2382.24 | 2498.31 | +4.87%  |
  | scheduler_pod_scheduling | SchedulingBasic/5000Nodes | Perc50   | ms     | 2447.23 | 2604.43 | +6.42%  |
  | scheduler_pod_scheduling | SchedulingBasic/5000Nodes | Perc90   | ms     | 4565.14 | 4613.83 | +1.07%  |
  | scheduler_pod_scheduling | SchedulingBasic/5000Nodes | Perc99   | ms     | 5064.51 | 5069.38 | +0.10%  |
  +--------------------------+---------------------------+----------+--------+---------+---------+---------+
  ```
</details>

<details>
  <summary>Preemption/5000Nodes</summary>
  
  ```
  +--------------------------+----------------------+----------+--------+-----------+-----------+--------+
  |          METRIC          |        GROUP         | QUANTILE |  UNIT  |    OLD    |    NEW    |  DIFF  |
  +--------------------------+----------------------+----------+--------+-----------+-----------+--------+
  | SchedulingThroughput     | Preemption/5000Nodes | Average  | pods/s |      9.90 |      9.78 | -1.18% |
  | SchedulingThroughput     | Preemption/5000Nodes | Perc50   | pods/s |      0.00 |      0.00 | NaN%   |
  | SchedulingThroughput     | Preemption/5000Nodes | Perc90   | pods/s |     42.40 |     42.20 | -0.47% |
  | SchedulingThroughput     | Preemption/5000Nodes | Perc99   | pods/s |     67.40 |     66.62 | -1.15% |
  | scheduler_e2e_scheduling | Preemption/5000Nodes | Average  | ms     |     36.57 |     36.53 | -0.10% |
  | scheduler_e2e_scheduling | Preemption/5000Nodes | Perc50   | ms     |     26.96 |     27.08 | +0.44% |
  | scheduler_e2e_scheduling | Preemption/5000Nodes | Perc90   | ms     |     85.95 |     84.09 | -2.16% |
  | scheduler_e2e_scheduling | Preemption/5000Nodes | Perc99   | ms     |    194.85 |    189.65 | -2.67% |
  | scheduler_pod_scheduling | Preemption/5000Nodes | Average  | ms     | 456669.70 | 460623.34 | +0.87% |
  | scheduler_pod_scheduling | Preemption/5000Nodes | Perc50   | ms     | 490881.59 | 491069.79 | +0.04% |
  | scheduler_pod_scheduling | Preemption/5000Nodes | Perc90   | ms     | 622464.32 | 622501.96 | +0.01% |
  | scheduler_pod_scheduling | Preemption/5000Nodes | Perc99   | ms     | 652070.43 | 652074.20 | +0.00% |
  +--------------------------+----------------------+----------+--------+-----------+-----------+--------+
  ```
</details>

<details>
  <summary>SchedulingNodeAffinity/5000Nodes</summary>
  
  ```
  +--------------------------+----------------------------------+----------+--------+---------+---------+---------+
  |          METRIC          |              GROUP               | QUANTILE |  UNIT  |   OLD   |   NEW   |  DIFF   |
  +--------------------------+----------------------------------+----------+--------+---------+---------+---------+
  | SchedulingThroughput     | SchedulingNodeAffinity/5000Nodes | Average  | pods/s |  125.00 |  125.00 | +0.00%  |
  | SchedulingThroughput     | SchedulingNodeAffinity/5000Nodes | Perc50   | pods/s |  117.00 |  120.00 | +2.56%  |
  | SchedulingThroughput     | SchedulingNodeAffinity/5000Nodes | Perc90   | pods/s |  190.50 |  187.40 | -1.63%  |
  | SchedulingThroughput     | SchedulingNodeAffinity/5000Nodes | Perc99   | pods/s |  190.50 |  187.40 | -1.63%  |
  | scheduler_e2e_scheduling | SchedulingNodeAffinity/5000Nodes | Average  | ms     |   91.10 |   77.56 | -14.86% |
  | scheduler_e2e_scheduling | SchedulingNodeAffinity/5000Nodes | Perc50   | ms     |   15.11 |   15.00 | -0.73%  |
  | scheduler_e2e_scheduling | SchedulingNodeAffinity/5000Nodes | Perc90   | ms     |   67.56 |   57.52 | -14.86% |
  | scheduler_e2e_scheduling | SchedulingNodeAffinity/5000Nodes | Perc99   | ms     | 2483.95 | 1880.22 | -24.31% |
  | scheduler_pod_scheduling | SchedulingNodeAffinity/5000Nodes | Average  | ms     | 3058.69 | 3080.09 | +0.70%  |
  | scheduler_pod_scheduling | SchedulingNodeAffinity/5000Nodes | Perc50   | ms     | 3297.63 | 3227.48 | -2.13%  |
  | scheduler_pod_scheduling | SchedulingNodeAffinity/5000Nodes | Perc90   | ms     | 4762.16 | 4757.87 | -0.09%  |
  | scheduler_pod_scheduling | SchedulingNodeAffinity/5000Nodes | Perc99   | ms     | 6228.38 | 6765.56 | +8.62%  |
  +--------------------------+----------------------------------+----------+--------+---------+---------+---------+
  ```
</details>

<details>
  <summary>SchedulingPodAffinity/5000Nodes</summary>
  
  ```
  +--------------------------+---------------------------------+----------+--------+----------+----------+--------+
  |          METRIC          |              GROUP              | QUANTILE |  UNIT  |   OLD    |   NEW    |  DIFF  |
  +--------------------------+---------------------------------+----------+--------+----------+----------+--------+
  | SchedulingThroughput     | SchedulingPodAffinity/5000Nodes | Average  | pods/s |    33.65 |    33.23 | -1.25% |
  | SchedulingThroughput     | SchedulingPodAffinity/5000Nodes | Perc50   | pods/s |    34.10 |    33.22 | -2.57% |
  | SchedulingThroughput     | SchedulingPodAffinity/5000Nodes | Perc90   | pods/s |    42.90 |    42.50 | -0.93% |
  | SchedulingThroughput     | SchedulingPodAffinity/5000Nodes | Perc99   | pods/s |    46.60 |    45.10 | -3.22% |
  | scheduler_e2e_scheduling | SchedulingPodAffinity/5000Nodes | Average  | ms     |    45.07 |    45.56 | +1.09% |
  | scheduler_e2e_scheduling | SchedulingPodAffinity/5000Nodes | Perc50   | ms     |    39.93 |    39.74 | -0.48% |
  | scheduler_e2e_scheduling | SchedulingPodAffinity/5000Nodes | Perc90   | ms     |    92.52 |    90.29 | -2.40% |
  | scheduler_e2e_scheduling | SchedulingPodAffinity/5000Nodes | Perc99   | ms     |   206.22 |   205.90 | -0.16% |
  | scheduler_pod_scheduling | SchedulingPodAffinity/5000Nodes | Average  | ms     | 14263.98 | 14465.57 | +1.41% |
  | scheduler_pod_scheduling | SchedulingPodAffinity/5000Nodes | Perc50   | ms     | 14118.47 | 14445.44 | +2.32% |
  | scheduler_pod_scheduling | SchedulingPodAffinity/5000Nodes | Perc90   | ms     | 33113.01 | 33451.37 | +1.02% |
  | scheduler_pod_scheduling | SchedulingPodAffinity/5000Nodes | Perc99   | ms     | 40175.30 | 40209.14 | +0.08% |
  +--------------------------+---------------------------------+----------+--------+----------+----------+--------+
  ```
</details>

<details>
  <summary>SchedulingPodAntiAffinity/5000Nodes</summary>
  
  ```
  +--------------------------+-------------------------------------+----------+--------+---------+---------+---------+
  |          METRIC          |                GROUP                | QUANTILE |  UNIT  |   OLD   |   NEW   |  DIFF   |
  +--------------------------+-------------------------------------+----------+--------+---------+---------+---------+
  | SchedulingThroughput     | SchedulingPodAntiAffinity/5000Nodes | Average  | pods/s |  118.00 |  117.57 | -0.37%  |
  | SchedulingThroughput     | SchedulingPodAntiAffinity/5000Nodes | Perc50   | pods/s |  115.60 |  111.40 | -3.63%  |
  | SchedulingThroughput     | SchedulingPodAntiAffinity/5000Nodes | Perc90   | pods/s |  178.90 |  178.00 | -0.50%  |
  | SchedulingThroughput     | SchedulingPodAntiAffinity/5000Nodes | Perc99   | pods/s |  178.90 |  178.00 | -0.50%  |
  | scheduler_e2e_scheduling | SchedulingPodAntiAffinity/5000Nodes | Average  | ms     |   88.81 |  131.50 | +48.07% |
  | scheduler_e2e_scheduling | SchedulingPodAntiAffinity/5000Nodes | Perc50   | ms     |   13.04 |   13.54 | +3.83%  |
  | scheduler_e2e_scheduling | SchedulingPodAntiAffinity/5000Nodes | Perc90   | ms     |  119.43 |  152.76 | +27.91% |
  | scheduler_e2e_scheduling | SchedulingPodAntiAffinity/5000Nodes | Perc99   | ms     | 2038.12 | 2649.58 | +30.00% |
  | scheduler_pod_scheduling | SchedulingPodAntiAffinity/5000Nodes | Average  | ms     | 1028.14 | 1061.80 | +3.27%  |
  | scheduler_pod_scheduling | SchedulingPodAntiAffinity/5000Nodes | Perc50   | ms     | 1028.81 | 1064.57 | +3.48%  |
  | scheduler_pod_scheduling | SchedulingPodAntiAffinity/5000Nodes | Perc90   | ms     | 2225.85 | 2267.39 | +1.87%  |
  | scheduler_pod_scheduling | SchedulingPodAntiAffinity/5000Nodes | Perc99   | ms     | 3165.23 | 3599.39 | +13.72% |
  +--------------------------+-------------------------------------+----------+--------+---------+---------+---------+
  ```
</details>

<details>
  <summary>TopologySpreading/5000Nodes</summary>
  
  ```
  +--------------------------+-----------------------------+----------+--------+----------+----------+--------+
  |          METRIC          |            GROUP            | QUANTILE |  UNIT  |   OLD    |   NEW    |  DIFF  |
  +--------------------------+-----------------------------+----------+--------+----------+----------+--------+
  | SchedulingThroughput     | TopologySpreading/5000Nodes | Average  | pods/s |    64.75 |    63.56 | -1.84% |
  | SchedulingThroughput     | TopologySpreading/5000Nodes | Perc50   | pods/s |    69.50 |    67.60 | -2.73% |
  | SchedulingThroughput     | TopologySpreading/5000Nodes | Perc90   | pods/s |    80.90 |    79.90 | -1.24% |
  | SchedulingThroughput     | TopologySpreading/5000Nodes | Perc99   | pods/s |    86.80 |    85.50 | -1.50% |
  | scheduler_e2e_scheduling | TopologySpreading/5000Nodes | Average  | ms     |    26.64 |    27.61 | +3.64% |
  | scheduler_e2e_scheduling | TopologySpreading/5000Nodes | Perc50   | ms     |    18.13 |    19.30 | +6.45% |
  | scheduler_e2e_scheduling | TopologySpreading/5000Nodes | Perc90   | ms     |    46.88 |    47.63 | +1.60% |
  | scheduler_e2e_scheduling | TopologySpreading/5000Nodes | Perc99   | ms     |   149.39 |   153.74 | +2.91% |
  | scheduler_pod_scheduling | TopologySpreading/5000Nodes | Average  | ms     | 14399.95 | 14615.29 | +1.50% |
  | scheduler_pod_scheduling | TopologySpreading/5000Nodes | Perc50   | ms     | 14435.52 | 14594.17 | +1.10% |
  | scheduler_pod_scheduling | TopologySpreading/5000Nodes | Perc90   | ms     | 33046.80 | 33264.58 | +0.66% |
  | scheduler_pod_scheduling | TopologySpreading/5000Nodes | Perc99   | ms     | 40168.68 | 40190.46 | +0.05% |
  +--------------------------+-----------------------------+----------+--------+----------+----------+--------+
  ```
</details>

#### Does this PR introduce a user-facing change?

```release-note
Scheduler's CycleState now embeds internal read/write locking inside its Read() and Write() functions. Meanwhile, Lock() and Unlock() function are removed.

action required: scheduler plugin developers are now required to remove CycleState#Lock() and CycleState#Unlock(). Just simply use Read() and Write() as they're natively thread-safe now.
```

Huang-Wei commented:
/hold

I don't expect the introduced read lock to introduce significant overhead. Will run perf tests to verify it.

Huang-Wei commented:
Updated the test result in `Special notes for your reviewer:` section.

/cc @alculquicondor @ahg-g 
FYI @fabiokung

fabiokung commented:
Thank you for pushing this forward @Huang-Wei! 👏

alculquicondor commented:
> I didn't see an obvious performance downgrade according to the bench diff:

```
| scheduler_pod_scheduling | SchedulingBasic/5000Nodes | Average  | ms     |  778.43 |  887.20 | +13.97% |
| scheduler_pod_scheduling | SchedulingBasic/5000Nodes | Perc50   | ms     |  524.26 |  814.97 | +55.45% |
```

That is very significant.

Huang-Wei commented:
> That is very significant.

To ensure it's not an outlier. I will rerun the test for `SchedulingBasic` for more rounds.

Huang-Wei commented:
Rerun the SchedulingBasic for 20 times, and updated the test result. @alculquicondor

ahg-g commented:
The results are quite noisy, but I am not concerned about performance, rlock shouldn't be very light weight (assuming no contention).
/lgtm 
/hold

alculquicondor commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:
```diff
@@ -298,6 +298,11 @@ func (pl *InterPodAffinity) RemovePod(ctx context.Context, cycleState *framework
+	cycleState.RLock()
+	defer cycleState.RUnlock()
+
 	c, err := cycleState.Read(preFilterStateKey)
```
Why not add the lock in the `Read` function?

ahg-g replied:
I was wondering the same, but I think the issue here is that adding the rlock inside Read syncs access to the map only, not the value.

alculquicondor replied:
but this is not about locking the value.

ahg-g replied:
True. Adding RLock to Read but not lock in the Write function is strange.

Huang-Wei replied:
> Why not add the lock in the `Read` function?

An enforced read lock inside `Read` is inappropriate as users should have the option to choose what kind of lock around `Read` - either simply read lock like the in-tree plugins, or write lock as the original issue stated.

ahg-g replied:
I would add a function to cycleState with a lock. Like `ReadWithRLock`?

alculquicondor replied:
> An enforced read lock inside `Read` is inappropriate as users should have the option to choose what kind of lock around `Read`

But there's no choice. We don't need the lock in the in-tree plugins, but we have to have it to avoid breaking out-of-tree plugins.

ahg-g replied:
The following will cause a deadlock:

```
cycleState.Lock()
c, err := cycleState.Read(key)
update(c)
cycleState.Write(key, c)
```

ahg-g replied:
not that this is something anyone should do...

Huang-Wei replied:
@ahg-g is right, adding an RLock inside would cause double locking - which is an anti-pattern in Go, although I'm not 100% sure it will cause deadlock. Here is what I want to achieve:

- Read pattern
    - explicit read lock around `Read()`
    - read it and process
- Write pattern
    - explicit write lock around `Read()`
    - mutate it during processing

Instead of:

- Write pattern
    - explicit write lock around `Read()`, but there is a read lock under the hood.

ahg-g replied:
Wei, the point is that the example I gave shouldn't be deployed at all, the lock should be acquired when accessing the map only, not the values. So what plugins should be doing is:

```
cycleState.RLock()
c, err := cycleState.Read(key)
cycleState.RUnLock()

// use a state specific lock if needed.
c = update(c)

cycleState.Lock()
cycleState.Write(key, c)
cycleState.UnLock()
```

The cycleState lock shouldn't be acquired to synchronize updating the value, it should be acquired only when reading/writing to the map.

ahg-g replied:
so locking inside Read and Write makes sense. But I don't want to add a read lock to Read and leave Write as is without locking. So we either adding explicit function that indicates a lock is acquired, or we add locking to both, which I think should also have no material impact on performance for in-tree plugins.

alculquicondor replied:
Actually, it seems that the place where we potentially have the most number of uses of `Read` is during preemption. Did you check those tests @Huang-Wei ?

If the increase is significant, we can have separate `Read` and `RLockRead` functions. And then we get rid of the `Lock` functions, as there shouldn't be anything using that lock for any other purpose.

Huang-Wei replied:
> the lock should be acquired when accessing the map only, not the values.

Yes, that's true. And that leads to a "best practice" - which is: modifying the map by introducing a "placeholder" key/value pair during PreFilter/PreScore, and then use **plugin-specific lock** in Filter/Score to mutate that **value**.

> Wei, the point is that the example I gave shouldn't be deployed at all,

We cannot guarantee everyone is aware and technically following the pattern. If users directly mutating the **map** during Filter/Score, which is running in parallel, the scheduler would panic. That's why we're introducing read locks in in-tree plugins, isn't it?

> The cycleState lock shouldn't be acquired to synchronize updating the value

It's what [volume_binding#Filter()](https://github.com/kubernetes/kubernetes/blob/0644c7f413f5c5985eec97378477ac4cfec0614a/pkg/scheduler/framework/plugins/volumebinding/volume_binding.go#L222-L224) is doing right now, which adheres to the best practice described above.

But again, what if a plugin doesn't follow this pattern and directly mutating the map?

>  So we either add an explicit function that indicates a lock is acquired (ReadWithRLock), or we add locking to both Read (RLock) and Write (Lock), which I think should have no material impact on performance for in-tree plugins.

Given that PreFilter/PreScore plugins are running in parallel and in-tree plugins only perform Write() in those two phases, I'm reluctant to add locking to Write(), which is pointless and would definitely add overhead.

---

So my point is: if users are adopting the "best practice" above, (basically) all they need is just Read lock during Filter/Score, accompanied with plugin-specific locks. In this case, we even don't need this PR.

But if users don't follow the "best practice", i.e., mutating the map in Filter/Score, what they do is read the map first, and write the value to the map. So lock-free Read() and Write() plus write locks around it would make more sense; instead of (implicit) read lock plus write lock.

> So what plugins should be doing is:
> 
> ```
> cycleState.RLock()
> c, err := cycleState.Read(key)
> cycleState.RUnLock()
> 
> // use a state specific lock if needed.
> c = update(c)
> 
> cycleState.Lock()
> cycleState.Write(key, c)
> cycleState.UnLock()
> ```

If a user has the `c` in place, they don't need to write it back b/c `c` is a pointer already:

```
cycleState.RLock()
c, err := cycleState.Read(key)
cycleState.RUnLock()

// use a state specific lock if needed.
c = update(c)
```

alculquicondor replied:
OTOH, if we don't have the locks in `Read`, any out-of-tree plugin could break another out-of-tree plugin.

So I would say: we either don't have this PR or we add the lock inside the Read function. I'm also in favor of Abdullah's suggestion of 2 methods: one with locking and one without.

Huang-Wei replied:
> we either don't have this PR or we add the lock inside the Read function.

The "either" part is correct: if everyone follows the "good practice", technically we don't need this PR. The "or" part is "conditionally" correct - when users don't program the plugin using the anti-pattern described above (read and also write to map in a Filter/Score plugin).

> I'm also in favor of Abdullah's suggestion of 2 methods: one with locking and one without.

I'm fine with simplifying this PR by introducing a `SafeRead()` function. `Read()` may not be needed anymore, but it's good to keep it as out-of-tree plugins may be using it. In terms of `SafeWrite()`, it is not needed for now but maybe it's harmless to create it. WDYT?

ahg-g replied:
> But again, what if a plugin doesn't follow this pattern and directly mutating the map?

Plugins can't mutate the map itself directly, it is a private member and can only be accessed via Read/Write funcs. Any synchronization needed on the value itself should be done via an external lock specific to that piece of data. Using the CycleState lock makes no sense.

> which is pointless and would definitely add overhead
It is not pointless if we want to provide proper threadsafe access to the map, and it doesn't add overhead if cycleState is properly used: plugins shouldn't be writing a lot to the cycleState map itself, each plugin should do it once per scheduling cycle, and that causes near zero locking overhead. 

The reason I want to have locking done inside both read and write funcs is because I want to remove all Lock related functions, they are confusing and will likely be misused and cause problems.

> If a user has the c in place, they don't need to write it back b/c c is a pointer already:

Sure, the point is that cycleState.Lock is mutex on accessing the map, not the values themselves, which are always references to an opaque data type.

Huang-Wei replied:
@ahg-g I think we're on the same page of how to efficiently use CycleState. What we're talking about here is how to deal with misuse of CycleState, from the perspective of defensive programming.

> because I want to remove all Lock related functions, they are confusing and will likely be misused and cause problems.

This is fair. So @alculquicondor @ahg-g what's your take on this:

- keep 4 functions: Read/SafeRead/Write/SafeWrite, and remove Lock/Unlock
    - in terms of sequential phases (PreFilter/PreScore), use Write or SafeWrite?
- keep 2 functions: Read/Write, and remove Lock/Unlock - i.e., we only have locking version of read/write.

alculquicondor replied:
I say, let's start with option 2 and evaluate performance. Please make sure preemption performance is not affected as much. I think we would lock for each Pod that is added/removed.

ahg-g replied:
sounds good, we also need to Lock inside `Delete` (not sure if anyone uses it).

Huang-Wei replied:
SG, will update and run the perf test.

Huang-Wei replied:
Updated the PR and testing result.

Note that the value of `scheduler_e2e_scheduling` varies a lot due to unpredictable binding duration. In my test, I tried my best to exclude the obvious outliers, but it's still challenging for this metric.

Huang-Wei committed:
sched: make CycleState's Read()/Write()/Delete() thread-safe

- add internal locking to CycleState's Read()/Write()/Delete() functions
- remove Lock() and Unlock() functions

alculquicondor commented:
Awesome, I'm happy with the results and the code change. I'll leave the LGTM to @ahg-g

ahg-g commented:
/lgtm

Huang-Wei commented:
/hold cancel

BinacsLee commented:
/retest

lining2020x commented:
@Huang-Wei @ahg-g Hello, will this MR be backported to v1.19? If so, when will this be done?

lining2020x commented:
> @Huang-Wei @ahg-g Hello, will this MR be backported to v1.19? If so, when will this be done?

 I have found the [cherry pick guide](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md)  😂  . With the help of `hack/cherry_pick_pull.sh`, I have just submmited a cherry-pick PR for v1.19 branch.

lining2020x commented:
As well as  PRs for branch release-1.20 and release-1.21.

alculquicondor commented:
/remove-kind bug
/kind feature

Unfortunately, I don't think this qualifies for backport: it's not a user facing bug.

Huang-Wei commented:
@lining2020x I don't quite think it's qualified for backporting. One reason is as @alculquicondor mentioned, it won't impact the end-users; the other reason is for out-of-tree plugin developers, you still have workarounds: e.g., inject a top-level state object as a placeholder, and then compose the locking logic inside the top-level state object in your plugin.

lining2020x commented:
@Huang-Wei I think this is a developer-friendly enhancement to the scheduler framework. According to the above test results of the performance impact caused by the increase of read-write locks, it did not bring obvious negative effects.
